Week-1/assignment1/readme.md

```diff
@@ -0,0 +1,60 @@
+ASSIGNMENT-1
+============
+
+### A - What is the main functionality of the browser?
+
+Explanation – The main functionality of browser is to read the requested HTML-document and render it into the webpage, to render HTML-document into webpage browser go through the sequence of processes as follows.
+
+![browser work flow](images/broserworkflow.png)
+
+### B - High Level Components of a browser.
+
+Explanation B – Browser’s higher order components are
+
+1.  The user interface: this includes the “address bar”, back/forward button, bookmarking menu, etc.
+2.  The browser engine: marshals actions between the UI and the rendering engine.
+3.  The rendering engine : responsible for displaying requested content. For example if the requested content is HTML, the rendering engine parses HTML and CSS, and displays the parsed content on the screen.
+4.  Networking: for network calls such as HTTP requests, using different implementations for different platform behind a platform-independent interface.
+5.  UI backend: used for drawing basic widgets like combo boxes and windows.
+6.  JavaScript interpreter. Used to parse and execute JavaScript code.
+7.  Data storage. This is a persistence layer. The browser may need to save all sorts of data locally, such as cookies.
+
+![highlevelComponent](images/highlevelComponent.png)
+
+### C - Rendering engine and its use.
+
+Explanation C - The responsibility of the rendering engine is well... Rendering, that is display of the requested contents on the browser screen. Rendering engine use: - By default, the rendering engine can display HTML and XML documents and images. It can display other types of data via plug-ins or extension; for example, displaying PDF documents using a PDF viewer plug-in. However, in this chapter we will focus on the main use case: displaying HTML and images that are formatted using CS
+
+![renderingEngine](images/renderingEngine.png)
+
+### D - Parsers (HTML, CSS, etc)
+
+Explanation D - a parser is a program, usually part of a compiler, that receives input in the form of sequential source program instructions, interactive online commands, markup tags, or some other defined interface and breaks them up into parts that can be used by other components in programming. Parsers usually divide the work between two components: the lexer (sometimes called tokenizer) that is responsible for breaking the input into valid tokens, and the parser that is responsible for constructing the parse tree by analyzing the document structure according to the language syntax rules. The lexer knows how to strip irrelevant characters like white spaces and line breaks.
+
```

Comment:
HTML Parser and CSS Parser each build a type of 'tree'. Please mention them here for the sake of completeness